### PR TITLE
Multiple ambig files

### LINKED ISSUE #557: HADDOCK3 lost the support for PDB-specific restraint files
In HADDOCK2.4, we can provide manually restraints files specific for each model to be calculated in it0. This requires putting the `ambig.tbl_XXX` files manually into the `structures/it0` directory (where XXX is the model number).

This mechanism is used by the server in case more than 50% random removal is turned on.

In HADDOCK3, although the CNS machinery does support it, we have no way to do it. I guess we don't want to specify in the cfg file e.g. 10000 such files. And when restarting a specific stage the directory is deleted, i.e. there is no way to manually copy the files into the specific directory...

We should think of a way of supporting this.  May-be providing the files as a tar archive (`ambig.tbl.tgz`) and in that case the python machinery should copy it to the proper directory and unpack it.

And there is more to it, as in HADDOCK2.4, the files are then copied to the next stage (it1), making sure the correct file is copied since there is a model renumbering happening between it0 and it1, which must also be the case in haddock3.

### PR DESCRIPTION
**Pull request by mgiulini**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and you comply with the following criteria:

- [x] You have stick to Python. Talk with us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [ ] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there's a (state) purpose
- [x] code follows our coding style
- [x] You wrote tests for the new code
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [x] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any *install dependencies* unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

Closes #557 by supporting the insertion of an `archive_name.tbl.tgz` file as the `ambig_fname` containing many (but at least one) tbl files. The syntax assumes that the names of the tbl files in the archive are consistent with `archive_name_0.tbl`, `archive_name_1.tbl`, etc. in order to avoid loading other tbl files (such as the `unambig_fname`). The tbl filename is appended to the cns input file in the `prepare_cns_input` function.

If the set of tbl files has been used by some CNS module, the option 'previous', added to each CNS module, allows one to re-use the same restraint file that was used previously. If no tbl is selected and previous is set to false, no ambiguous restraints are employed.

If the number of tbl files is lower than the number of models to be generated, the tbls are evenly distributed over the whole set. If the number is higher, the files in excess are simply ignored.

Scoring modules only deliver the restraint information contained in the pdbs to the next module, without using it in the cns execution.

### CODE REVIEW

**Comment by mgiulini** [comment]
TODO list / list of questions:

- [ ] scoring modules cancel the `restr_fname` attribute of PDBFiles, as there's no `ambig_fname` in them. Should scoring modules retain information about restraints? One pro is that this would preserve modularity, but one con is that it might be misleading.
- [ ] `ambig_fname` keyword is repeated in the cns input file. `ambig_fname = ambig.tbl.tgz` would result in an input file with the following lines
```
eval ($ambig_fname="../data/1_rigidbody/ambig.tbl.tgz")
eval ($ambig_fname="../data/1_rigidbody/ambig_3.tbl")
```
while a more standard `ambig_fname = ambig.tbl.tgz` would give
```
eval ($ambig_fname="../data/1_rigidbody/ambig.tbl")
eval ($ambig_fname="../data/1_rigidbody/ambig.tbl")
```
I have no problems with this repetition, as the second line always overwrites the first.
- [ ] implement tests
- [ ] gdock is left untouched
- [ ] `previous` option is present also in rigidbody, but in that case it is not used, as it doesn't make sense. Should I remove it?

File: src/haddock/modules/refinement/flexref/defaults.yaml [hunk 1]
```diff
@@ -30,6 +30,16 @@ ambig_fname:
     Note that this file can contain both ambiguous and unambiguous restraints.
   group: 'distance restraints'
   explevel: easy
+previous:
+  default: false
+  type: boolean
+  title: Boolean condition on the ambiguous distance restraints file
+  short: Boolean condition on the ambiguous distance restraints file. 
+    If False, do not use any restraints. If True, use the the restraints coming from the previous CNS module.
+  long: Boolean condition on the ambiguous distance restraints file.
```
[the remark this answers is not in the document]
Comment (amjjbonvin) [reply src/haddock/modules/refinement/flexref/defaults.yaml:39]: And same comment for other modules.

[the remark this answers is not in the document]
Comment (amjjbonvin) [reply src/haddock/modules/refinement/flexref/defaults.yaml:39]: I would think that setting previous to false should trigger the use of the restraint file defined for this particular step if present.

[the remark this answers is not in the document]
Comment (mgiulini) [reply src/haddock/modules/refinement/flexref/defaults.yaml:39]: Yes, this is the way the code is implemented: when `previous = false` (default behaviour) and there exists an `ambig_fname`, the latter is used. The description is indeed a bit misleading, as it refers to the case when no `ambig_fname` is provided. I'll fix it.

[the remark this answers is not in the document]
Comment (mgiulini) [reply src/haddock/modules/refinement/flexref/defaults.yaml:39]: the presence of an `ambig_fname` overwrites any restraint information coming from the previous stages: if `previous = true` and `ambig_fname` is defined, the latter is used as well.

**Comment by amjjbonvin** [comment]
May-be `previous` should be renamed to `previous-ambig` or `previous-tbl` for clarity?

**Comment by mgiulini** [comment]
> May-be `previous` should be renamed to `previous-ambig` or `previous-tbl` for clarity?

definitely!

File: src/haddock/modules/refinement/flexref/defaults.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgiulini) [reply src/haddock/modules/refinement/flexref/defaults.yaml:39]: done!

**Review by amjjbonvin — APPROVED** [review]

**Pull request merged**